### ISSUE TITLE
Wallet notify not working

### ISSUE THREAD

**Issue by snomannaseem** [bitcoin/bitcoin#12090]
Hi, 

I am getting following error against wallet notify event.

runCommand error: system(/home/ec2-user/cron/receive.sh 1a8ae3367556b76bcf660a748d2a1c01c37ec10a931b48c50e963b4115397e2b) returned -1

- I tried lots of options in configuration file. Some of them are listed as follows
walletnotify="/home/ec2-user/cron/receive.php %s"
walletnotify=/home/ec2-user/cron/receive.php "%s"
walletnotify="/home/ec2-user/cron/receive.php" %s
walletnotify=/bin/bash /home/ec2-user/cron/receive.sh %s
walletnotify=/bin/php /home/ec2-user/cron/receive.php %s

- I also tried by creating a separate page and configured it through CURL command as follows
   walletnotify=/bin/curl http://127.0.0.1/receive.php?trx=%s

I also tried by running the **bitciond** daemon from root user. I also verified folder permissions which are fine.

In short I have been struggling with this from couple of days but unable to succeed.
I always get same error which is **return -1** along with the command that I entered in configuration file.

Not to mention that rest of the **bitcoin core** features are working fine. My block chain database is also up to date.

Here are my current configuration parameters.

----------------------------------------------------------------------------

testnet=1
server=1
rpcuser=root
rpcpassword=123
rpcallowip=127.0.0.1
rpcport=6806
walletnotify=/home/ec2-user/cron/receive.sh %s

----------------------------------------------------------------------------

Is there something I am missing?

**Comment by bolekC** [comment]
Looks like configuration error, can you attach more log lines and list detailed /home/ec2-user/cron/receive.sh permissions? Configuration looks OK for me. Error message shows like correct command was executed. And if you can attach also receive.sh file.

**Comment by bolekC** [comment]
After the night I have an idea, modify the line:
walletnotify=**/bin/bash** /home/ec2-user/cron/receive.sh %s

System call does not work directly for shell scripts. It expects binary file so you have to call shell directly.

**Comment by snomannaseem** [comment]
Thanks for your reply,

I have changed the walletnotify value as above but still getting the same error.
I have attached a file.zip folder which contains following files.

1) Daemon log file.
2) Configuration file.
3) Directory snapshot for permissions.
4) receive.sh file.

In the log file please search for the following transaction id you will find that the error is still appearing 

b82057b45b663acb04f3e6b9fd304c8e1dfa16b2eb32a1a4b74f25c7c5b3fbd0

[files.zip](https://github.com/bitcoin/bitcoin/files/1606654/files.zip)

**Comment by bolekC** [comment]
I tried to reproduce this. Copied your config and `receive.sh` script. (only changing directories to have mine)

It works fine. 

The only way to get `runCommand error: system....` messages was to modify bitcoin.conf to have wrong path to `receive.sh` file. So check if the filepath to `/home/ec2-user/cron/receive.sh` is OK.

**Comment by snomannaseem** [comment]
If path is incorrect why it runs fine when I copy and paste the same command line to cli from error message?

**Comment by bolekC** [comment]
I've tried this also on amazon could as you do. Works fine. Same file names, same directories. No problem.

Looks like you are running bitcoind as ec2-user, the same one as all receive.sh so I have no idea what is wrong. It does not reproduce on my servers. 

Have you tried on other servers? 
Other sysops? I've trided on Debian 7 and Amazon standard linux. No problems found so its not bitcoind config problem. 

In my opinion you have correct bitcoin.conf file.

**Comment by snomannaseem** [comment]
All above experiments were done with CentOS-7 (64bit) on my local VMWare machine and Red Hat Linux on Amazon EC2 server but got the same error.

By the way here is another post stating similar issue https://bitcointalk.org/index.php?topic=385146.0 
Looks like I am missing something which is related to some default configuration of operating system. 

Anyways today I tried it with Debian 7 on my local machine here is what happened.

1. I generated a new address through bitcoin-cli.
2. Against the new address I tried to issue free bitcoins from one of the testnet faucet.
3. I can see transaction number on the faucet website which shows transaction was created successfully on faucet website.
4. But I don't see any wallet notify event in debug.log file. Please correct me If I am guessing wrong that one of the possible reason could be the db is in synchronization process with the main block chain testnet database. Once synchronization is completed then it will issue walletnotify command.

Is there any other way I can trigger this walletnotify event seprately for testing purpose?

**Comment by MarcoFalke** [comment]
Please let us know if

./test <https://github.com/bitcoin/bitcoin/tree/master/test>/functional
<https://github.com/bitcoin/bitcoin/tree/master/test/functional>/
*notifications.py*


*passes.*

**Comment by snomannaseem** [comment]
One more thing........ I download bitcoin core installation package from following URL.

https://bitcoin.org/bin/bitcoin-core-0.15.1/bitcoin-0.15.1-x86_64-linux-gnu.tar.gz 

I un-tared it and ran the bitcoin daemon.

I haven't compiled the source on my machine.

**Comment by MarcoFalke** [comment]
Ah, in case you are using the compiled binaries:

Download the targz with the sources (including tests). And run the p2p-versionbits test, which uses "alertnotify" instead of walletnotify. But they use the same method `runCommand()`.

```sh
bitcoin-0.15.1$ ./test/functional/p2p-versionbits-warning.py --srcdir=./bin/folder/where/your/bitcoind/is/
```

**Comment by bolekC** [comment]
You should get notification after synchronization is finished. I've had the same. They cam later.

CentOs and Redhat are almost the same. It might be something with related to sysop but I have no clue now what that can be. I had no time today to try on one of these systems. Maybe tomorrow.

You can try the test as suggested but I guess they will fail.

**Comment by bolekC** [comment]
If you still have a problem can you do some tracing? I'm not getting this on RedHat.

Please run bitcoind with stracing command:
`strace -f -o bit.log ./bitcoind -daemon &`

Then reproduce the fault.

Please then attach bit.log file so we can check what is going one there. Maybe that will help to troubleshoot.

**Comment by snomannaseem** [comment]
1. Ok, I am able to run walletnotify on Debian virtual machine and also on the old CentOS-7 VM. I only moved bitcoind service to root folder and the receive.sh file as well and gave root permissions to it and it is running fine.

2. BUT I am still getting the same error with Amazon EC2 machine which is running Red Hat on it. Though I executed the same steps but still getting the error. Is there any different security settings in RedHat linux.

3. Also to switch user to root amazon only requires **sudo su** command no password. Is there any difference between normal root user and amazon root user from security point of view? 

4. Also I am attaching bit.log file which I created from following command on Amazon Red Hat Server.
 strace -f -o bit.log ./bitcoind -daemon &

[error-in-debug.log.txt](https://github.com/bitcoin/bitcoin/files/1612903/error-in-debug.log.txt)
[bit.zip](https://github.com/bitcoin/bitcoin/files/1612921/bit.zip)

**Comment by bolekC** [comment]
I've looked at the log, it looks like you are running out of memory. Bitcoind uses a lot of memory so check it. I'm getting strange problem running bitcoind on a small instances in Amazon EC2 so try it on a bigger instance. Which EC2 instance you are using?

The error is from the thread that prints just immediately the error in debug log.
`4331  clone(child_stack=0, flags=CLONE_PARENT_SETTID|SIGCHLD, parent_tidptr=0x7fdabd218ad0) = -1 ENOMEM (Cannot allocate memory)`

Hope that will help.

**Comment by snomannaseem** [comment]
Ok, finally the problem is resolved. 

You were right the issue was related to low memory. My VM had 1 GB RAM with no swap which I thought was enough but after your previous post I added 4 GB of swap and then tried again and it worked fine. 

Thanks again for your help and prompt replies.

**Comment by ishankoodles** [comment]
I used both ways of using walletnotify but the curl command is getting hits thrice when sending the bitcoin within the app, i m looking for an solution for last 2 days
